File: java/components/StrataPersistence/src/subsystems/strata1/entity/hibernaterepository/MoneyType.java
// ##########################################################################
// # File Name:	MoneyType.java
// #
// # Copyright:	2012, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.entity.hibernaterepository;

import strata1.common.money.Money;
import org.hibernate.HibernateException;
import org.hibernate.engine.spi.SessionImplementor;
import org.hibernate.type.CurrencyType;
import org.hibernate.type.DoubleType;
import org.hibernate.type.Type;
import org.hibernate.usertype.CompositeUserType;
import java.io.Serializable;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Currency;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class MoneyType 
    implements CompositeUserType
{
    /************************************************************************
     * Creates a new {@code DateTimeType}. 
     *
     */
    public
    MoneyType() {}
    
    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public String[] 
    getPropertyNames()
    {
        return new String[] {"Currency","Amount"};
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Type[] 
    getPropertyTypes()
    {
        return new Type[] { new CurrencyType(),new DoubleType() };
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Object 
    getPropertyValue(Object component,int property)
        throws HibernateException
    {
        Money money = (Money)component;
        
        if ( property == 0) 
            return money.getCurrency();
        
        if ( property == 1 )
            return new Double(money.getAmount());
        
        throw new HibernateException("unknown property");
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    setPropertyValue(Object component,int property,Object value)
        throws HibernateException
    {
        throw new HibernateException("Money is immutable.");
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Class<?> 
    returnedClass()
    {
        return Money.class;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    equals(Object x,Object y) 
        throws HibernateException
    {
        if ( x == y )
            return true;
        
        if ( x == null || y == null )
            return false;
        
        return x.equals( y );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public int 
    hashCode(Object x) 
        throws HibernateException
    {
        if ( x == null )
            return 0;
        
        return x.hashCode();
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Object 
    nullSafeGet(
        ResultSet          rs,
        String[]           names,
        SessionImplementor session,
        Object             owner) 
            throws 
                HibernateException,
                SQLException
    {
        Currency currency = null;
        double   amount   = 0.0;
        
        currency = Currency.getInstance(rs.getString(names[0]));
        
        if ( rs.wasNull() )
            return null;
        
        amount = rs.getDouble( names[1] );
        
        if ( rs.wasNull() )
            return null;
        
        return new Money(currency,amount);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    nullSafeSet(
        PreparedStatement  st,
        Object             value,
        int                index,
        SessionImplementor session) 
            throws 
                HibernateException,
                SQLException
    {
        if ( value == null )
        {
            st.setNull( index,new CurrencyType().sqlType() );
            st.setNull( index+1,new DoubleType().sqlType() );
        }
        else
        {
            st.setString(index,((Money)value).getCurrency().getCurrencyCode());
            st.setDouble(index+1,((Money)value).getAmount());
        }
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Object 
    deepCopy(Object value) 
        throws HibernateException
    {
        return (Money)value;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isMutable()
    {
        return false;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Serializable 
    disassemble(
        Object             value,
        SessionImplementor session)
            throws HibernateException
    {
        return (Money)value;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Object 
    assemble(
        Serializable       cached,
        SessionImplementor session,
        Object             owner) 
            throws HibernateException
    {
        return (Money)cached;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public Object 
    replace(
        Object             original,
        Object             target,
        SessionImplementor session,
        Object             owner) 
            throws HibernateException
    {
        return (Money)original;
    }

}

// ##########################################################################

File: csharp/components/Strata.Nhibernate/Mapping/MoneyType.cs
//  ##########################################################################
//  # File Name: MoneyType.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using NHibernate;
using NHibernate.Engine;
using NHibernate.Type;
using NHibernate.UserTypes;
using Strata.Foundation.Value;
using System;
using System.Data.Common;

namespace Strata.Nhibernate.Mapping
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// User defined type Nhibernate mapper for <c>Money</c> objects.
    /// </summary>
    /// 
    public 
    class MoneyType:
        ICompositeUserType
    {
        public String[] 
        PropertyNames
        {
            get { return new String[] {"Currency","Amount"}; }
        }

        public IType[] 
        PropertyTypes
        {
            get { return new IType[] {NHibernateUtil.String,NHibernateUtil.Decimal}; }
        }

        public Type 
        ReturnedClass
        {
            get { return typeof(Money); }
        }

        public bool
        IsMutable
        {
            get { return false; }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>MoneyType</c> instance.
        /// </summary>
        /// 
        public 
        MoneyType() {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType.GetPropertyValue(Object,int)"/>
        /// </summary>
        /// 
        public Object 
        GetPropertyValue(Object component,int property)
        {
            Money money = (Money)component;

            if (money == null)
                return null;

            if ( property == 0 ) 
                return money.Currency.CurrencyCode;
        
            if ( property == 1 )
                return money.Amount;
        
            throw new HibernateException("unknown property");
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see 
        /// cref="ICompositeUserType.SetPropertyValue(Object,int,Object)"/>
        /// </summary>
        /// 
        public void 
        SetPropertyValue(Object component,int property,Object value)
        {
            //throw new HibernateException("Money is immutable.");
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType.Equals(Object,Object)"/>
        /// </summary>
        /// 
        public new bool
        Equals(Object x,Object y) 
        {
            if (ReferenceEquals(x,y))
                return true;
        
            if ( x == null || y == null )
                return false;
        
            return x.Equals( y );
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType.GetHashCode(Object)"/>
        /// </summary>
        /// 
        public int 
        GetHashCode(Object x) 
        {
            if ( x == null )
                return 0;
        
            return x.GetHashCode();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Object 
        NullSafeGet(
            DbDataReader        reader,
            string[]            names,
            ISessionImplementor session,
            object              owner) 
        {
            string  currencyColumn = names[0];
            string  amountColumn   = names[1];
            object  currency       = null;
            object  amount         = 0.0m;

            if( reader == null )
                return null;
            
            currency = 
                NHibernateUtil
                    .String
                    .NullSafeGet(
                        reader,
                        currencyColumn,
                        session,
                        owner);
            amount = 
                NHibernateUtil
                    .Decimal
                    .NullSafeGet(
                        reader,
                        amountColumn,
                        session,
                        owner);

            if (currency == null || amount == null)
                return null;

            return 
                new Money(
                    CurrencyManager.GetInstanceByCode( currency.ToString() ),
                    (decimal)amount);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public void 
        NullSafeSet(
            DbCommand           st,
            Object              value,
            int                 index,
            bool[]              settable,
            ISessionImplementor session) 
        {
            String  currency = "";
            Decimal amount = 0.0m;

            if (value == null)
            {
                NHibernateUtil.String.NullSafeSet(st,null,index,session);
                NHibernateUtil.Decimal.NullSafeSet(st,null,index + 1,session);
                return;
            }

            amount = ((Money) value).Amount;
            currency = ((Money) value).Currency.CurrencyCode;
 
            NHibernateUtil.String.NullSafeSet(st,currency,index,session);
            NHibernateUtil.Decimal.NullSafeSet(st,amount,index+1,session);
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType.DeepCopy(Object)"/>
        /// </summary>
        /// 
        public Object 
        DeepCopy(Object value) 
        {
            return (Money)value;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType
        ///                .Disassemble(Object,ISessionImplementor)"/>
        /// </summary>
        /// 
        public Object 
        Disassemble(
            Object              value,
            ISessionImplementor session)
        {
            return (Money)value;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType
        ///                .Assemble(Object,ISessionImplementor,Object)"/>
        /// </summary>
        /// 
        public Object 
        Assemble(
            Object              cached,
            ISessionImplementor session,
            Object              owner) 
        {
            return (Money)cached;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="ICompositeUserType
        ///                .Replace(
        ///                    Object,Object,ISessionImplementor,Object)"/>
        /// </summary>
        /// 
        public Object 
        Replace(
            Object              original,
            Object              target,
            ISessionImplementor session,
            Object              owner) 
        {
            return (Money)original;
        }

    }
}

//  ##########################################################################
